### PR TITLE
[TreeProcMT] Add support for chain of trees with different names

### PR DESCRIPTION
**Pull request by eguiraud**
Internal logic has been adapted to support the case in which TTreeProcessorMT
is acting on a TChain and the trees in the TChain do not all have the same name.
Importantly, the fTreeName data-member changed name to fTreeNames and is now a
vector of strings rather than a single string.

This fixes ROOT-10178.

### CODE REVIEW

File: tree/treeplayer/src/TTreeProcessorMT.cxx [hunk 1]
```diff
@@ -43,17 +43,17 @@ struct EntryCluster {
 
 ////////////////////////////////////////////////////////////////////////////////
 /// Construct fChain, also adding friends if needed and injecting knowledge of offsets if available.
-void TTreeView::MakeChain(const std::string &treeName, const std::vector<std::string> &fileNames,
+void TTreeView::MakeChain(const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames,
```
Comment (hageboeck) [inline tree/treeplayer/src/TTreeProcessorMT.cxx:46]:
```suggestion
/// \param[in] treeNames Name of the tree for each file in `fileNames`.
/// \param[in] fileNames Files to be opened.
/// \param[in] friendInfo xxx No idea what this is.
/// \param[in] nEntries Number of entries to be processed.
/// \param[in] friendEntries xxx Why do we need that? 
void TTreeView::MakeChain(const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames,
```

File: tree/treeplayer/src/TTreeProcessorMT.cxx [hunk 2]
```diff
@@ -158,7 +158,8 @@ TTreeView::GetTreeReader(Long64_t start, Long64_t end, const std::string &treeNa
 /// Return a vector of cluster boundaries for the given tree and files.
 // EntryClusters and number of entries per file
 using ClustersAndEntries = std::pair<std::vector<std::vector<EntryCluster>>, std::vector<Long64_t>>;
-static ClustersAndEntries MakeClusters(const std::string &treeName, const std::vector<std::string> &fileNames)
+static ClustersAndEntries
+MakeClusters(const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames)
```
Comment (hageboeck) [inline tree/treeplayer/src/TTreeProcessorMT.cxx:162]: Doxygen probably doesn't parse that correctly. Suggest moving the `using` before the `///` comment.

File: tree/treeplayer/src/TTreeProcessorMT.cxx (hunk 1 above)
Comment (hageboeck) [reply tree/treeplayer/src/TTreeProcessorMT.cxx:46]: Maybe it's just an internal function, but certainly doesn't do any harm.

Comment (eguiraud) [reply tree/treeplayer/src/TTreeProcessorMT.cxx:51]: Yes, TTreeView is an internal helper class. Thanks though, committed!

File: tree/treeplayer/src/TTreeProcessorMT.cxx (hunk 2 above)
Comment (eguiraud) [reply tree/treeplayer/src/TTreeProcessorMT.cxx:167]: Doxygen should parse that correctly...?

Comment (eguiraud) [reply tree/treeplayer/src/TTreeProcessorMT.cxx]: Fixed

File: tree/treeplayer/src/TTreeProcessorMT.cxx [hunk 3]
```diff
@@ -43,17 +43,22 @@ struct EntryCluster {
 
 ////////////////////////////////////////////////////////////////////////////////
 /// Construct fChain, also adding friends if needed and injecting knowledge of offsets if available.
-void TTreeView::MakeChain(const std::string &treeName, const std::vector<std::string> &fileNames,
+/// \param[in] treeNames Name of the tree for each file in `fileNames`.
+/// \param[in] fileNames Files to be opened.
+/// \param[in] friendInfo xxx No idea what this is.
+/// \param[in] nEntries Number of entries to be processed.
+/// \param[in] friendEntries xxx Why do we need that?
```
Comment (hageboeck) [inline tree/treeplayer/src/TTreeProcessorMT.cxx:50]: Ah, well. The github suggestions don't really work so well if the suggestion is incomplete ...

If you look closely, you see that I had no clue what `friendInfo` and `friendEntries` are supposed to do. That's why I started the docstring.

Comment (eguiraud) [reply tree/treeplayer/src/TTreeProcessorMT.cxx:50]: oh boy xD

Comment (eguiraud) [reply tree/treeplayer/src/TTreeProcessorMT.cxx:50]: fixed

File: tree/treeplayer/src/TTreeProcessorMT.cxx [hunk 4]
```diff
@@ -271,16 +280,21 @@ GetFriendEntries(const std::vector<std::pair<std::string, std::string>> &friendN
...
+      auto &chain = static_cast<const TChain &>(tree);
       auto files = chain.GetListOfFiles();
-      if (files && 0 != files->GetEntries()) {
-         return files->At(0)->GetName();
+      if (!files || files->GetEntries() == 0) {
+         throw std::runtime_error("TTreeProcessorMT: input TChain does not contain any file");
       }
+      std::vector<std::string> treeNames;
+      for (TObject *f : *files)
+         treeNames.emplace_back(f->GetName());
```
Comment (etejedor) [inline tree/treeplayer/src/TTreeProcessorMT.cxx:295]: So when you do `GetName` on a TFile that contains exactly one tree, it returns the name of the tree?

Comment (eguiraud) [reply tree/treeplayer/src/TTreeProcessorMT.cxx:295]: It's not a `TFile` thing, it's a "file-coming-from-TChain" thing. Line 286 was supposed to clarify this.

Comment (etejedor) [reply tree/treeplayer/src/TTreeProcessorMT.cxx:295]: So it is the `TChain` that sets the names of its files to be the names of the trees each file contains (which form the chain)?

File: tree/treeplayer/src/TTreeProcessorMT.cxx [hunk 5]
```diff
@@ -365,37 +379,41 @@ Internal::FriendInfo TTreeProcessorMT::GetFriendInfo(TTree &tree)
 
 ////////////////////////////////////////////////////////////////////////////////
 /// Retrieve the name of the first TTree in the first input file, else throw.
```
Comment (etejedor) [inline tree/treeplayer/src/TTreeProcessorMT.cxx:381]: This needs to be something like "Retrieve the names of each TTree in each of the input files, else throw."

File: tree/treeplayer/src/TTreeProcessorMT.cxx (hunk 4 above)
Comment (etejedor) [reply tree/treeplayer/src/TTreeProcessorMT.cxx:295]: Ok thanks. I would do a slight rewording of the comment: `// Case 1: this is a TChain: for each file it contains, GetName returns the name of the tree in that file`.

File: tree/treeplayer/src/TTreeProcessorMT.cxx (hunk 5 above)
Comment (eguiraud) [reply tree/treeplayer/src/TTreeProcessorMT.cxx]: fixed

File: tree/treeplayer/src/TTreeProcessorMT.cxx (hunk 4 above)
Comment (eguiraud) [reply tree/treeplayer/src/TTreeProcessorMT.cxx:295]: fixed

**Pull request merged**